File: 100. Same Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        bool res = true;
        preorder(p, q, res);
        return res;
    }
    
    void preorder(TreeNode* p, TreeNode* q, bool &res)
    {
        if (p == NULL && q == NULL)
            return;
        else if (p != NULL && q == NULL)
        {
            res = false;
            return;
        }
        else if (p == NULL && q != NULL)
        {
            res = false;
            return;
        }
        if (p->val != q->val)
            res = false;
        preorder(p->left, q->left, res);
        preorder(p->right, q->right, res);
    }
};

File: 100. Same Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        self.res = True
        self.preorder(p, q)
        return self.res
    
    def preorder(self, p, q):
        if p == None and q == None:
            return
        elif p == None and q:
            self.res = False
            return
        elif p and q == None:
            self.res = False
            return
        if p.val != q.val:
            self.res = False
        self.preorder(p.left, q.left)
        self.preorder(p.right, q.right)
